codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttEncoder.java

```diff
@@ -234,23 +274,38 @@ private static ByteBuf encodeSubscribeMessage(
         // Variable Header
         int messageId = variableHeader.messageId();
         buf.writeShort(messageId);
+        buf.writeBytes(propertiesBuf);
 
         // Payload
         for (MqttTopicSubscription topic : payload.topicSubscriptions()) {
-            String topicName = topic.topicName();
-            byte[] topicNameBytes = encodeStringUtf8(topicName);
-            buf.writeShort(topicNameBytes.length);
-            buf.writeBytes(topicNameBytes, 0, topicNameBytes.length);
-            buf.writeByte(topic.qualityOfService().value());
+            writeUTF8String(buf, topic.topicName());
+            final MqttSubscriptionOption option = topic.option();
+
+            int optionEncoded =  option.retainHandling().value() << 4;
+            if (option.isRetainAsPublished()) {
+                optionEncoded |= 0x08;
+            }
+            if (option.isNoLocal()) {
+                optionEncoded |= 0x04;
+            }
+            optionEncoded |= option.qos().value();
+
+            buf.writeByte(optionEncoded);
         }
 
         return buf;
     }
 
     private static ByteBuf encodeUnsubscribeMessage(
+            ChannelHandlerContext ctx,
             ByteBufAllocator byteBufAllocator,
             MqttUnsubscribeMessage message) {
-        int variableHeaderBufferSize = 2;
+        MqttVersion mqttVersion = MqttCodecUtil.getMqttVersion(ctx);
+        ByteBuf propertiesBuf = encodePropertiesIfNeeded(mqttVersion,
```

Comment:
you will need to call `propertiesBuf.release()` later as you not transfer ownership